src/devices/Common/Iot/Device/Common/ValueArray.cs

```diff
@@ -211,15 +226,15 @@ public void Insert(int index, T item)
             }
 
             int i = _count - 1;
-            var self = AsSpanFull();
+
+            _count++; // Increase count first, so we can access [i + 1]
```

Comment:
You can do `Unsafe.Add(ref _e0, i)` rather than incrementing the count, which will work and prevent potentially exposing the temporary length change to callers in multi-threaded scenarios.

Reply:
Good idea, but the class is not thread safe in either case (the other way round, it would probably be possible to insert more than the maximum number of elements in a parallel loop)

Reply:
That is a fair point. I must have skimmed and not realized that this was code in an `Add`/`Insert` method to begin with.